src/eve/pattern_matching.py

```diff
@@ -0,0 +1,159 @@
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+import inspect
+from dataclasses import dataclass
+from functools import singledispatch
+from types import ModuleType
+from typing import Any, Iterator
+
+
+@dataclass(frozen=True)
+class ObjectPattern:
+    """Class to pattern match general objects.
+
+    A pattern matches an object if it is an instance of the specified
+    class and all attributes of the pattern (recursively) match the
+    objects attributes.
+
+    >>> class Foo:
+    ...    def __init__(self):
+    ...        self.bar = 1
+    ...        self.baz = 2
+    >>> assert ObjectPattern(Foo, {"bar": 1}).match(Foo())
+    """
+
+    cls: type
+    attrs: dict[str, Any]
+
+    def match(self, other: Any, raise_exception: bool = False) -> bool:
+        """Return if object pattern matches `other` using :func:`get_differences`.
+
+        If `raise_exception` is specified raises an exception with all differences
+        found.
+        """
+        if raise_exception:
+            diffs = [*get_differences(self, other)]
+            if len(diffs) > 0:
+                diffs_str = "\n  ".join(
+                    [f"  {self.cls.__name__}{path}: {msg}" for path, msg in diffs]
+                )
+                raise ValueError(f"Object and pattern don't match:\n  {diffs_str}")
+            return True
+
+        return next(get_differences(self, other), None) is None
+
+    def __str__(self) -> str:
+        attrs_str = ", ".join([f"{str(k)}={str(v)}" for k, v in self.attrs.items()])
+        return f"{self.cls.__name__}({attrs_str})"
+
+
+@dataclass(frozen=True)
+class _ObjectPatternConstructor:
+    """Helper class to construct an ObjectPattern.
+
+    This is just an explicit way of doing partial function application
+    and was choosen to improve debuggability.
+    """
+
+    cls: type
+
+    def __call__(self, **kwargs: Any) -> ObjectPattern:
+        return ObjectPattern(self.cls, kwargs)
+
+
+@singledispatch
+def get_differences(a: Any, b: Any, path: str = "") -> Iterator[tuple[str, str]]:
+    """Compare two objects and return a list of differences.
+
+    If the arguments are lists or dictionaries comparison is recursively per item. Objects are compared
+    using equality operator or if the left-hand-side is an `ObjectPattern` its type and attributes
+    are compared to the right-hand-side object. Only the attributes of the `ObjectPattern` are used
+    for comparison, disregarding potential additional attributes of the right-hand-side.
+    """
+    if type(a) != type(b):
```

Comment:
Types are singletons, so this could be a better way to check:

```suggestion
    if type(a) is not type(b):
```